Implement in-memory GSD/FSD validation to avoid disk I/O

jpanasiukTT opened the pull request:
### Ticket
No ticket, problem described in https://github.com/tenstorrent/tt-metal/pull/31341#issuecomment-3569266500
Also issue: https://github.com/tenstorrent/tt-metal/issues/33062

### Problem description
We had unnecessary disk I/O during GSD/FSD validation by working with in-memory representations throughout the validation pipeline.

### What's changed
- Add in-memory overload of validate_fsd_against_gsd() that accepts protobuf and YAML::Node objects directly
- Refactor get_factory_system_descriptor_path() to get_factory_system_descriptor() which returns FSD protobuf object instead of writing/reading from disk
- Update validate_connectivity() to use in-memory objects (YAML::Node, FSD proto)
- Update handle_workload_timeout() to use PhysicalSystemDescriptor::generate_yaml_node() and in-memory FSD generation via CablingGenerator::generate_factory_system_descriptor()
- Update run_connectivity_validation() in run_cluster_validation.cpp to use in-memory validation, eliminating need for temp file cleanup
- Remove cleanup_metadata() function as it's no longer needed

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

tt-asaigal reviewed (approved):
Looks good - cleans up a lot of things giving us a headache.

jpanasiukTT commented:
Rerunning
apc: https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml
multi host: https://github.com/tenstorrent/tt-metal/actions/runs/19685055701
t3k: https://github.com/tenstorrent/tt-metal/actions/runs/19685057603

Commit:
Add in-memory validation for cluster connectivity

- Implement in-memory FSD validation against GSD without file I/O
- Add get_factory_system_descriptor() function for in-memory proto generation
- Update cluster validation to use in-memory validation approach
- Remove file-based validation dependencies

Commit:
Merge main and add documentation/validation fixes

- Merge origin/main into branch
- Add documentation for in-memory validate_fsd_against_gsd() overload
- Add input validation to get_factory_system_descriptor() to prevent bad_optional_access
- Improve error messages for missing parameters

tt-asaigal commented on tools/scaleout/validation/utils/cluster_validation_utils.cpp:
```diff
@@ -1534,22 +1534,58 @@ std::string get_factory_system_descriptor_path(
+fsd::proto::FactorySystemDescriptor get_factory_system_descriptor(
+    const std::optional<std::string>& cabling_descriptor_path,
+    const std::optional<std::string>& deployment_descriptor_path,
+    const std::optional<std::string>& fsd_path) {
```
Why are we not passing in the hostname vector here. I see this function being logically identical to the one that emits the FSD to disk

tt-asaigal reviewed (changes requested):
seeing some issues after latest changes were made

jpanasiukTT committed:
review
